lib/qiita/markdown/filters/external_link.rb

```diff
@@ -0,0 +1,45 @@
+require "addressable/uri"
+
+module Qiita
+  module Markdown
+    module Filters
+      class ExternalLink < HTML::Pipeline::Filter
+        def call
+          doc.search("a").each do |anchor|
+            next unless anchor["href"]
+            href = anchor["href"].strip
+            href_host = host_of(href)
+            next unless href_host
+            if external_link?(href_host)
+              anchor["rel"] = "nofollow"
+              anchor["target"] = "_blank"
+            end
+          end
+
+          doc
+        end
+
+        def validate
+          needs :hostname
+        end
+
+        private
+
+        def host_of(url)
+          uri = Addressable::URI.parse(url)
+          uri.host
+        rescue Addressable::URI::InvalidURIError
+          nil
+        end
+
+        def hostname
+          context[:hostname]
+        end
+
+        def external_link?(target_host)
+          target_host != hostname && !target_host.end_with?(".#{hostname}")
+        end
```

Comment:
IMO a method with the name `#external_link?` should take a URL as an argument instead of a host (i.e. invoke `#host_of` in `#external_link?`).